### ISSUE TITLE
Questions about the Stack-LLaMa

### ISSUE THREAD

**Issue by beyondguo** [lvwerra/trl#492]
Hi, I see the https://github.com/lvwerra/trl/tree/main/examples/stack_llama/scripts and found it a good RLHF tutorial. However, there are some steps I can't figure out.

The first step is "Supervised fine-tuning of the base `llama-7b` model to create `llama-7b-se`", and the second step is "Reward modeling using dialog pairs from the SE dataset using the `llama-7b-se` to create `llama-7b-se-rm`".

Why do you train the reward model based on `llama-7b-se`, rather than `llama-7b`?

And I see the model loading scripts(from https://github.com/lvwerra/trl/blob/main/examples/stack_llama/scripts/reward_modeling.py) is :
```python
model = AutoModelForSequenceClassification.from_pretrained(
    script_args.model_name, num_labels=1, torch_dtype=torch.bfloat16
) 
model = get_peft_model(model, peft_config)
model.print_trainable_parameters()
```
the `model_name` seems to be the `llama-7b-se` path, which is a LoRA weights path, is this the right way the load a LoRA model? And does this mean that in the reward modeling step, a new LoRA model is added to the first SFT LoRA model?

I'm not sure if I put it clearly, lol. Looking forward to your reply, many thanks!

**Comment by lvwerra** [comment]
> Why do you train the reward model based on llama-7b-se, rather than llama-7b?

In practice, it helps if the base model was already fine-tuned on the target data, such as `llama-7b-se`, so the reward model gets a bit better.

> the model_name seems to be the llama-7b-se path, which is a LoRA weights path, is this the right way the load a LoRA model? And does this mean that in the reward modeling step, a new LoRA model is added to the first SFT LoRA model?

We merged the adapters into the base model before the RM training with the [merging script](https://github.com/lvwerra/trl/blob/main/examples/stack_llama/scripts/merge_peft_adapter.py). Then the training script simply adds a new adapter on top of that model

Hope this helps!

**Comment by beyondguo** [comment]
Got it! Thanks!

**Comment by lvwerra** [comment]
Closing this as the questions seems answered - feel free to re-open if not!

**Comment by beyondguo** [comment]
Hi @lvwerra , I'm tuning the PPO training these days but found the training is quite unstable. Could you please give some advice of the parameter setting?

My problem is that after several iterations, the model seems start to generate empty outputs which however have high rewards:
![image](https://github.com/lvwerra/trl/assets/37113676/48bd03d1-3bba-4e4d-aa13-b3722889f6de)

And this seems to lead to an error: `autodetected range of [nan, nan] is not finite`.

I guess there're some hyper-params that I can tune to aviod this behavior, could you give my suggestions? Thanks!

**Comment by halfrot** [comment]
> Hi @lvwerra , I'm tuning the PPO training these days but found the training is quite unstable. Could you please give some advice of the parameter setting?
> 
> My problem is that after several iterations, the model seems start to generate empty outputs which however have high rewards: ![image](https://user-images.githubusercontent.com/37113676/253136476-48bd03d1-3bba-4e4d-aa13-b3722889f6de.png)
> 
> And this seems to lead to an error: `autodetected range of [nan, nan] is not finite`.
> 
> I guess there're some hyper-params that I can tune to aviod this behavior, could you give my suggestions? Thanks!

Having the same issue `autodetected range of [nan, nan] is not finite`. I've tried some combinations of `mini_batch_size` and `gradient_accumulation_steps` according to https://github.com/huggingface/trl/issues/545#issue-1812590848, but I still can't figure out a solution.
@lvwerra Could you please help with some advice? Thank you.

**Comment by lvwerra** [comment]
Maybe you should add a term to the reward function that penalises short responses? You can also force the `generate` function to at least produce `n` tokens.

**Comment by sunlibo2390** [comment]
> Maybe you should add a term to the reward function that penalises short responses? You can also force the `generate` function to at least produce `n` tokens.

So does this mean that short responses are easier to lead to something like gradient explosion causing the `nan` error?

**Comment by beyondguo** [comment]
I have a workaround: https://mp.weixin.qq.com/s/O5dfAfgNO7TudQIGqxg6Dw
which simply adds the EOS token to the `begin_suppress_tokens` parameter.

**Comment by lvwerra** [comment]
You can set the `eos_token_id=-1` in the generation kwargs to guarantee that the model doesn't stop generating even if an EOS token occurs. You can decide how to handle that case then in the reward function: maybe to compute the reward you cut-off everything after the eos token or you give a penalty to it.

**Comment by halfrot** [comment]
It appears that adjusting `min_new_tokens` to prevent short responses is effective. Thank you all!

**Comment by lvwerra** [comment]
Note that it can lead to negative KL issues so the suggested approach is safer with `max_new_tokens` and `eos_token_id=-1`. See https://huggingface.co/docs/trl/main/en/how_to_train#what-is-the-concern-with-negative-kl-divergence
